packages/@aws-cdk/aws-s3/lib/bucket.ts

```diff
@@ -480,7 +491,28 @@ abstract class BucketBase extends Resource implements IBucket {
   public urlForObject(key?: string): string {
     const stack = Stack.of(this);
     const prefix = `https://s3.${stack.region}.${stack.urlSuffix}/`;
-    return this.buildUrl(prefix, key);
+    if (typeof key !== 'string') {
+      return this.urlJoin(prefix, this.bucketName);
+    }
+    return this.urlJoin(prefix, this.bucketName, key);
+  }
+
+  /**
+   * The virtual-hosted style URL of an S3 object. For example:
+   * @example https://only-bucket.s3-us-west-1.amazonaws.com
+   * @example https://bucket.s3-us-west-1.amazonaws.com/key
+   * @example https://china-bucket.s3-cn-north-1.amazonaws.com.cn/mykey
+   * @param key The S3 key of the object. If not specified, the URL of the
+   *      bucket is returned.
+   * @returns an ObjectS3Url token
+   */
+  public virtualHostedUrlForObject(key?: string): string {
+    const stack = Stack.of(this);
+    const prefix = `https://${this.bucketName}.s3-${stack.region}.${stack.urlSuffix}`;
```

Comment:
```suggestion
    const prefix = `https://${this.bucketRegionalDomainName}`;
```

This will create `s3.<region>` instead of `s3-<region>` but thats ok, it also works.

Optionally, if you're up for it, lets also support building a non-regional URL like the [blog](https://aws.amazon.com/jp/blogs/aws/amazon-s3-path-deprecation-plan-the-rest-of-the-story/) mentions:

> `https://jbarr-public.s3.amazonaws.com/images/ritchie_and_thompson_pdp11.jpeg`

i.e something like this:

```ts
bucket.virtualHostedUrlForObject('my-key', { regional: false })
```

So basically use `this.bucketDomainName` if `regional` is `false` and `this.bucketRegionalDomainName` if `regional` is `true`.

Reply:
Wow, I didn't noticed about non-regional URL. I commited to support them :heart: Thank you!